packages/editor/src/components/provider/index.js

```diff
@@ -1,34 +1,56 @@
 /**
  * External dependencies
  */
-import { flow, map } from 'lodash';
+import { map } from 'lodash';
 
 /**
  * WordPress Dependencies
  */
 import { compose } from '@wordpress/compose';
-import { createElement, Component } from '@wordpress/element';
-import { DropZoneProvider, SlotFillProvider } from '@wordpress/components';
-import { withDispatch } from '@wordpress/data';
+import { Component } from '@wordpress/element';
+import { withDispatch, withSelect } from '@wordpress/data';
 import { __ } from '@wordpress/i18n';
+import { BlockEditorProvider } from '@wordpress/block-editor';
 
 /**
  * Internal dependencies
  */
 import { traverse, wrap, urlRewrite } from '../../editor-styles';
 
+/**
+ * Compute the Editor Provider's state object from props.
+ * @param {Object} props Component props.
+ *
+ * @return {Object} Component state.
+ */
+function computeProviderStateFromProps( props ) {
+	return {
+		settings: props.settings,
+		meta: props.meta,
+		onMetaChange: props.onMetaChange,
+		editorSettings: {
+			...props.settings,
+			__experimentalMetaSource: {
```

Comment:
This is a first iteration of what could be a generic way to declare custom attribute sources. I didn't want to push the proposal a lot. I think we should just land the minimal version here and work on it (polish it) as a separate PR.